csharp/src/Apache.Arrow/Column.cs

```diff
@@ -62,7 +62,16 @@ private bool ValidateArrayDataTypes()
         {
             for (int i = 0; i < Data.ArrayCount; i++)
             {
-                if (Data.Array(i).Data.DataType != Field.DataType)
+                if (Data.Array(i).Data.DataType.TypeId != Field.DataType.TypeId)
+                {
+                    return false;
+                }
+
+                var dataTypeComparer = new ArrayDataTypeComparer(Field.DataType);
```

Comment:
This can be lifted out of the `for` loop - there is no need to create a new one every time through the loop.

Reply:
Indeed!